### DISCUSSION TITLE
Worker type=module

### DISCUSSION THREAD

**Discussion by jasonsanjose** [webpack/webpack#17071] [Q&A] [answered]
I'm trying to understand how module workers should work with Webpack. 

Let's start with the example I found in the repo: https://github.com/webpack/webpack/blob/main/examples/worker/README.md

Looking at this example, there's a `new Worker(..., { type: "module" })` [link](https://github.com/webpack/webpack/blame/0e0dab0a4b75bf84fa4d9a974e6cc55f70486352/examples/worker/README.md#L85) in the source file but the compiled output has `new Worker(..., { type: undefined })` [link](https://github.com/webpack/webpack/blame/0e0dab0a4b75bf84fa4d9a974e6cc55f70486352/examples/worker/README.md#L482).

I'm trying to understand how to change the output behavior that emits the synchronous `importScripts()` instead of an async ESM `import()`.

For reference, I was basing my understanding of ESM in Workers and preloading based on this article: https://web.dev/module-workers/.

**Comment by alexander-akait** [comment]
You can't use `{ type: "module" }` and `importScripts` together without hacks, just build it without `{ type: "module" }`

**Reply by jasonsanjose** [reply]
Right. I don't want to use `importScripts`. 😄 

How do I stop webpack from doing that? Based on that example code, what is missing to make it emit an async `import()`?

**Reply by alexander-akait** [reply, marked as answer]
You can enable this using https://webpack.js.org/configuration/output/#outputworkerchunkloading and set it to `import`, but you need to enable ESM output using https://webpack.js.org/configuration/output/#outputmodule

**Reply by jasonsanjose** [reply]
Thank you! 🙏 

That worked perfectly. Here's the updated config:

```javascript
module.exports = {
	entry: "./example.js",
	output: {
		path: path.join(__dirname, "dist"),
		filename: "[name].js",
		chunkFilename: "[name].js",
		publicPath: "/dist/",
		workerChunkLoading: "import",
		module: true
	},
	optimization: {
		concatenateModules: true,
		usedExports: true,
		providedExports: true,
		chunkIds: "deterministic" // To keep filename consistent between different modes (for example building only)
	},
	experiments: {
		outputModule: true
	}
};
```

The output no longer contains `importScripts()`. I now see this in `dist/fibonacci.js`:

```javascript
var promise = import("../" + __webpack_require__.u(chunkId)).then(installChunk, (e) => {
	if(installedChunks[chunkId] !== 0) installedChunks[chunkId] = undefined;
	throw e;
});
```

**Reply by jasonsanjose** [reply]
I filed a follow up bug here https://github.com/webpack/webpack/issues/17125. I'm not able to get the same example to correctly `import()` a worker chunk at runtime.
